### ISSUE TITLE
This is crazy and useless, so much code (configuration, CONTROLLERS) for every entity

### ISSUE THREAD

**Issue by freerider7777** [OData/AspNetCoreOData#241]
There is PostgREST - you just start it and can use database as REST
https://github.com/PostgREST/postgrest

Is there a one-liner method like UseAllDbSets() - for quick export of all DBSets in DBContext?

**Comment by xuzhg** [comment]
@freerider7777 Do you mean use "DBContext" as starting point to build the Edm model automatically?

**Comment by freerider7777** [comment]
Yes, it was easy in ADO.Net Data Services (as far as I remember :) ), why make it complicated? Please have a look:
https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/cc907912(v=msdn.10)?redirectedfrom=MSDN

**Comment by robertmclaws** [comment]
You should be checking our Restier! (https://github.com/odata/restier). It's the successor to WCF Data Services and lets you use conventions to configure automatic, secure OData services.

There is some complexity in OData if you want more control over what happens. Otherwise, use Restier. No docs yet, but the samples should point you in the right direction. This is the latest version for ASP.NET Core: https://www.nuget.org/packages/Microsoft.Restier.AspNetCore/1.0.0-CI-20210719-021012

HTH!

**Comment by freerider7777** [comment]
Thanks!

**Comment by TehWardy** [comment]
What about something like this ...
```cs
using Microsoft.AspNetCore.OData;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.OData.ModelBuilder;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Api
{
    public static class IServiceCollectionExtensions
    {

        public static IServiceCollection AddDbsToOData(this IServiceCollection services, Type[] contextTypes)
        {
            ODataConventionModelBuilder builder = new();
            var models = new List<dynamic>();

            foreach (var ctxType in contextTypes)
            {
                var sets = ctxType.GetProperties().Where(p => p.PropertyType is IQueryable);
                var entitySet = builder.GetType().GetMethod("EntitySet");
                foreach (var set in sets)
                {
                    entitySet.MakeGenericMethod(set.PropertyType.GenericTypeArguments).Invoke(builder, new[] { set.PropertyType.Name });
                }

                models.Add(new { Path = ctxType.Name, Model = builder.GetEdmModel());
            }

            services.AddControllers()
                .AddOData(opt =>
                {
                    opt.EnableAttributeRouting = true;
                    opt.Expand().Count().Filter().Select().OrderBy().SetMaxTop(1000);
                    models.ForEach(m =>
                    {
                        opt.AddRouteComponents(m.Path, m.Model);
                    });
                });

            return services;
        }
   }
}
```

**Comment by freerider7777** [comment]
@TehWardy Thanks, but it seems it does not help without adding controllers...

**Comment by TehWardy** [comment]
Yeh you can implement a generic controller like I have .
I also don't do this as I don't want my entire DB exposed as part of the OData API, things like EF Migrations tables you often want to leave out, although I have an ApiIgnore attribute that I use to selective rip bits out.

Do you need help defining a generic controller too @freerider7777 ?

**Comment by TehWardy** [comment]
The only thing i'm not sure of is if I ONLY had something like this ...
```cs
public class GenericController<T> : ODataController { ... }
```
How would I tell the MVC framework how to map the requests in to it.

I do things a bit differently in that I have that but for each set I have explicitly implemented a derived type like ...
```cs
public class FooController : GenericController<Foo> { ... }
```

I have also marked my generic one as abstract so the framework won't attempt to map that.
The main reasoning for me is that exposing something on my API is done explicitly ALWAYS rather than just by convention.

This forces me to think about the implementation of each Set<T> that I expose fully before I have it in my API layer.

**Comment by freerider7777** [comment]
Sure there are different cases :) I just needed to transparently expose the database as web service for integration testing purposes. With a minimum effort :)
Thanks for the solution!
For now I use PostgREST, which works fine out of the box. I'll play around with GenericControllers later...

**Comment by JanKotschenreuther** [comment]
I have been waiting very long for a .NET Core compatible RESTier package to port our application to .NET Core. I thought that would not happen anymore.
A few months ago I ported our application to .NET 5 and had to come up with my own solution.
You can find that here https://github.com/OData/AspNetCoreOData/issues/98#issuecomment-859715535 .

**Comment by robertmclaws** [comment]
Jan, sorry to here that, but glad you came up with a solution! If you get a chance, you should check out the new Microsoft.Restier.AspNetCore package. Still a few outstanding issues we're working on, but we're in the home stretch of being able to ship the RTM.

**Comment by closem90** [comment]
Any help here yet?

**Comment by julealgon** [comment]
This should either be closed or moved to discussions.
